### PR TITLE
Autocomplete  progress bar

### PR DESCRIPTION
**Pull request by nightwing**
https://github.com/ajaxorg/ace/assets/341801/994ccaa8-cfed-40f4-9c29-2c70aa439a12

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by whazor** [comment]
For accessibility, would it be possible to use [`aria-busy="true"`](https://dockyard.com/blog/2020/03/02/accessible-loading-indicatorswith-no-extra-elements) to indicate that the autocomplete is not yet finished?

**Comment by akoreman** [comment]
> For accessibility, would it be possible to use [`aria-busy="true"`](https://dockyard.com/blog/2020/03/02/accessible-loading-indicatorswith-no-extra-elements) to indicate that the autocomplete is not yet finished?

If I'm not mistaken, `aria-busy` is there to tell screen readers to wait with reading (usually `aria-live`) content until `aria-busy` is set back to false. Not sure in this case we want to wait with reading the results until all completers are done?

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-busy

**Comment by nightwing** [comment]
@akoreman what should we show for prefers reduced motion case, the same animation but slower?

As for screen reader, how much delay for autocompleters do we expect? I'd expect in typical case loading state to take much shorter than what it would take screenreader to announce start and end of loading state, so i'd say not announcing is better than announcing exactly the way it is displayed.

**Comment by akoreman** [comment]
> what should we show for prefers reduced motion case, the same animation but slower?

The ideal would be to [completely eliminate motion](https://www.w3.org/WAI/WCAG21/Understanding/animation-from-interactions.html) when `prefers-reduced-motion` is enabled. Do you think a stationary version of this indicator would be clear to users that something is loading?

> As for screen reader, how much delay for autocompleters do we expect? I'd expect in typical case loading state to take much shorter than what it would take screenreader to announce start and end of loading state, so i'd say not announcing is better than announcing exactly the way it is displayed.

For some completers I could see it taking around 5-10 seconds to get results (e.g. AI based code suggestions), maybe we could read something to screen readers after e.g. 1 second, that should be slow enough to not be triggered by 'normal' completers but still offer an explanation for longer wait times?

**Comment by nightwing** [comment]
Added support for prefers-reduced-motion, since we need a separate code path for accessibility support, is it ok if i implement it in a separate pull request?

**Comment by akoreman** [comment]
> Added support for prefers-reduced-motion, since we need a separate code path for accessibility support, is it ok if i implement it in a separate pull request?

Yeah that sounds reasonable to me

File: src/autocomplete.js [hunk 1]
```diff
@@ -86,6 +86,19 @@ class Autocomplete {
         this.stickySelectionTimer = lang.delayedCall(function() {
             this.stickySelection = true;
         }.bind(this), this.stickySelectionDelay);
+
+        this.$firstOpenTimer = lang.delayedCall(function() {
+            var initialPosition = this.completionProvider && this.completionProvider.initialPosition;
+            if (this.autoShown || (this.popup && this.popup.isOpen) || !initialPosition) return;
+
+            var completionsForEmpty = [{
+                caption: "Loading...",
```
Comment (akoreman) [inline src/autocomplete.js:95]: We should allow this string to be translated through the `nls` function we added.

Comment (nightwing) [reply src/autocomplete.js]: done

**Review by akoreman — APPROVED** [review]

**Pull request merged**